R/utils-rows.R

```diff
@@ -0,0 +1,125 @@
+#' A data frame.
+#'
+#' @rdname rownames
+#'
+#' @examples
+#' # Convert between row names and column --------------------------------
+#' test <- rownames_as_column(mtcars, var = "car")
+#' test
+#' head(column_as_rownames(test, var = "car"))
+#'
+#' @export
+rownames_as_column <- function(x, var = "rowname") {
+  if (!insight::object_has_rownames(x)) {
+    insight::format_error("The data frame doesn't have rownames.")
+  }
+  if (is.null(var)) {
+    var <- "rowname"
+  }
+  if (!is.character(var)) {
+    insight::format_error("Argument 'var' must be of type character.")
+  }
+  if (var %in% colnames(x)) {
+    insight::format_error(
+      paste0("There is already a variable named `", var, "` in your dataset.")
+    )
+  }
+  original_x <- x
+  rn <- data.frame(rn = rownames(x), stringsAsFactors = FALSE)
+  x <- cbind(rn, x)
+  colnames(x)[1] <- var
+  rownames(x) <- NULL
+  x <- .replace_attrs(x, attributes(original_x))
+  x
+}
+
+#' @rdname rownames
+#' @export
+column_as_rownames <- function(x, var = "rowname") {
+  if (!is.character(var) && !is.numeric(var)) {
+    insight::format_error("Argument `var` must be of type character or numeric.")
+  }
+  if (is.character(var) && !var %in% names(x)) {
+    insight::format_error(paste0("Variable \"", var, "\" is not in the data frame."))
+  }
+  if (is.numeric(var) && (var > ncol(x) || var <= 0)) {
+    insight::format_error("Column ", var, " does not exist. There are ", ncol(x), " columns in the data frame.")
+  }
+
+  original_x <- x
+  rownames(x) <- x[[var]]
+  x[[var]] <- NULL
+  x <- .replace_attrs(x, attributes(original_x))
+  x
+}
+
+
+
+#' @rdname rownames
+#' @export
+#' @examples
+#' test_data <- head(iris)
+#'
+#' rowid_as_column(test_data)
+#' rowid_as_column(test_data, var = "my_id")
+rowid_as_column <- function(x, var = "rowid") {
+  UseMethod("rowid_as_column")
+}
+
+#' @export
+rowid_as_column.default <- function(x, var = "rowid") {
+  if (is.null(var)) {
+    var <- "rowid"
+  }
+  if (!is.character(var)) {
+    insight::format_error("Argument 'var' must be of type character.")
+  }
+  if (var %in% colnames(x)) {
+    insight::format_error(
+      paste0("There is already a variable named `", var, "` in your dataset.")
+    )
+  }
+  original_x <- x
+  rn <- data.frame(rn = seq_len(nrow(x)), stringsAsFactors = FALSE)
+  x <- cbind(rn, x)
+  colnames(x)[1] <- var
+  rownames(x) <- NULL
+  x <- .replace_attrs(x, attributes(original_x))
+  x
+}
+
+
+#' @export
+rowid_as_column.grouped_df <- function(x, var = "rowid") {
+
+  if (!is.character(var)) {
+    insight::format_error("Argument 'var' must be of type character.")
+  }
+  if (var %in% colnames(x)) {
+    insight::format_error(
+      paste0("There is already a variable named `", var, "` in your dataset.")
+    )
+  }
+  # works only for dplyr >= 0.8.0
+  grps <- attr(x, "groups", exact = TRUE)
+  grps <- grps[[".rows"]]
+
+  for (i in seq_len(length(grps))) {
+    x[grps[[i]], var] <- seq_len(length(grps[[i]]))
+  }
```

Comment:
I think you can replace both `seq_len(length())` with `seq_along()`.

Reply:
But I wonder if this is really what we want? We would have duplicated row id's here, i.e. each "group" starts with ID 1 to ...

Reply:
I thought `tibble::rowid_to_column()` did the same way as I do here for groups but apparently it just ignores them. I implemented this group behavior to fix the example in #416 but maybe I should remove it

Reply:
I'd still like to have an easy way to create row index per group that would be equivalent to:
``` r
library(dplyr, warn.conflicts = FALSE)

iris |> 
  sample_n(10) |> 
  group_by(Species) |> 
  mutate(id = row_number()) |> 
  ungroup()
#> # A tibble: 10 × 6
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species       id
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>      <int>
#>  1          5.8         2.8          5.1         2.4 virginica      1
#>  2          4.7         3.2          1.6         0.2 setosa         1
#>  3          6.4         2.9          4.3         1.3 versicolor     1
#>  4          6           2.7          5.1         1.6 versicolor     2
#>  5          7           3.2          4.7         1.4 versicolor     3
#>  6          6.3         2.8          5.1         1.5 virginica      2
#>  7          4.4         3            1.3         0.2 setosa         2
#>  8          7.7         2.8          6.7         2   virginica      3
#>  9          5.7         3            4.2         1.2 versicolor     4
#> 10          6.1         2.6          5.6         1.4 virginica      4
```
but I don't know if we want to implement a new function for this or keep it in a grouped method for `rowid_as_column()`

Reply:
We can keep this for the grouped_df method, I just wanted to be sure this is intentional.

Reply:
I added some details about this in the docs and in NEWS